File: C/quick_sort.c
#include <stdio.h>
#include <stdlib.h>
#include "functions.h"

int findPivotIndex(int *array, int start, int end){
	int pivot = start;
	start++;
	while(start <= end){
		if(array[pivot] > array[start]){
			start++;
		}else if(array[pivot] <= array[end]){
			end--;
		}else{
			swap(array, start, end);
			start++;
			end--;
		}
	}
	swap(array, pivot, end);
	return end;
}

void quickSort(int *array, int start, int end){
	if(start < end){
		int pivotIndex = findPivotIndex(array, start, end);
		quickSort(array, start, pivotIndex - 1);
		quickSort(array, pivotIndex + 1, end);
	}
}

int main(){
	int size = 20;
	int *array = createRandomArray(size);
	//printArray(array,20);
	quickSort(array, 0, size - 1);
	//printArray(array, 20);
	printf("Is array sorted: %d\n", isArraySorted(array, size));
	free(array);
}

File: Kotlin/quick_sort.kt
fun findPivotIndex(array:MutableList<Int>, start:Int, end:Int):Int{
	var pivotIndex = start
	var i = start + 1
	var j = end
	while(i <= j){
		if(array[pivotIndex] > array[i]) i++
		else if(array[pivotIndex] <= array[j]) j--
		else{
			var temp = array[i]
			array[i] = array[j]
			array[j] = temp
			i++
			j--
		}
	}
	var temp = array[pivotIndex]
	array[pivotIndex] = array[j]
	array[j] = temp
	return j
}

fun quickSort(array:MutableList<Int>, start:Int, end:Int){
	if(start < end){
		val pivotIndex = findPivotIndex(array, start, end)
		quickSort(array, start, pivotIndex - 1)
		quickSort(array, pivotIndex + 1, end)
	}
}

fun main(){
	val size = 20
	val array = MutableList(size){it + 1}
	array.shuffle()
	println(array.joinToString())
	quickSort(array,0,size-1)
	println(array.joinToString())
}



